wave_utils/src/pose_cov_comp.cpp

```diff
@@ -0,0 +1,514 @@
+
+    q = normalizeQuat(q);
+
+    return q;
+}
+
+Vector7 PoseWithCovariance::getPoseQuaternion() const {
+    Vector7 pose;
+    Vector4 q = this->getQuaternion();
+
+    pose << this->position, q;
+
+    return pose;
+}
+
+Eigen::Affine3d PoseWithCovariance::getTransformMatrix() const {
+    Eigen::Affine3d T_m;
+
+    T_m.translation() = this->position;
+    T_m.linear() = this->rotation_matrix;
+
+    return T_m;
+}
+
+PoseWithCovariance composePose(PoseWithCovariance &p1, PoseWithCovariance &p2) {
+    PoseWithCovariance r;  // store all the results
+
+    // Use Eigen Transform to transform poses
+    // This is the same as implementing Equation (5.5)
+    Eigen::Affine3d T_p1, T_p2, T_r;
+    T_p1 = p1.getTransformMatrix();
+    T_p2 = p2.getTransformMatrix();
+    T_r = T_p1 * T_p2;
+
+    r.position = T_r.translation();
+    r.rotation_matrix = T_r.rotation();
+
+    // compute the covariances
+    // p6 = [x y z roll pitch yaw]', transformation using YPR
+    // p7 = [x y z qr, qx, qy, qz]', transformation using Quaternion, qr is qw
+
+    // get covariances from PoseWithCovariance objects
+    Matrix6x6 cov_p1 = p1.covariance, cov_p2 = p2.covariance;
+
+    // pR7 is the composed pose in p7 form
+    // p17 is the p1 in p7 form, p27 is p2 in p7 form, p16 is p1 in p6 form,
+    // p26 is p2 in p6 form
+    Vector7 pR7, p17, p27;
+    Vector6 p16, p26;
+
+    pR7 = r.getPoseQuaternion();
+    p17 = p1.getPoseQuaternion();
+    p27 = p2.getPoseQuaternion();
+
+    p16 << p1.getPosition(), quatToYPR(p17.block<4, 1>(3, 0));
+    p26 << p2.getPosition(), quatToYPR(p27.block<4, 1>(3, 0));
+
+    // Equation (5.3)
+    Matrix6x7 jacobian_p7_to_p6 = jacobian_p7_to_p6_wrt_p(pR7);
+    Matrix7x7 jacobian_p7_p7_composition =
+      jacobian_p7_p7_Composition_wrt_p1(p17, p27);
+    Matrix7x6 jacobian_p6_to_p7 = jacobian_p6_to_p7_wrt_p(p16);
+    Matrix6x6 dfpc_dp =
+      jacobian_p7_to_p6 * jacobian_p7_p7_composition * jacobian_p6_to_p7;
+
+    // Equation (5.4)
+    jacobian_p7_p7_composition = jacobian_p7_p7_Composition_wrt_p2(p17, p27);
+    jacobian_p6_to_p7 = jacobian_p6_to_p7_wrt_p(p26);
+    Matrix6x6 dfpc_dq =
+      jacobian_p7_to_p6 * jacobian_p7_p7_composition * jacobian_p6_to_p7;
+
+    // Equation (5.2)
+    // putting everything together
+    Matrix6x6 final_cov = dfpc_dp * cov_p1 * dfpc_dp.transpose() +
+                          dfpc_dq * cov_p2 * dfpc_dq.transpose();
+
+    r.covariance = final_cov;
+
+    return r;
+}
+
+Vector4 normalizeQuat(const Vector4 q) {
+    Vector4 norm_q;
+    double norm_coeff;
+    double qr = q(0), qx = q(1), qy = q(2), qz = q(3);
+
+    norm_coeff = 1 / sqrt(qr * qr + qx * qx + qy * qy + qz * qz);
+    norm_q = norm_coeff * q;
+    return norm_q;
+}
+
+Vector3 quatToYPR(const Vector4 &q) {
+    Vector3 v;
+    double qr = q(0), qx = q(1), qy = q(2), qz = q(3);
+    double delta = qr * qy - qx * qz;
+    double roll, pitch, yaw;
+
+    if (fabs(delta - 0.5) < 1e-6) {
```

Comment:
Some comments here could be a little useful. Not exactly sure what the `delta` value itself represents, and why its value would lead to different use cases.

Reply:
Done